flatdata-py/flatdata/lib/archive_builder.py

```diff
@@ -0,0 +1,290 @@
+
+    @classmethod
+    def name(cls):
+        '''Returns archive name'''
+        return cls._NAME
+
+    @classmethod
+    def schema(cls):
+        '''Returns archive schema'''
+        return cls._SCHEMA
+
+    def _write_raw_data(self, name, data):
+        '''
+        Helper function to write data
+
+        :param name(str): resource name
+        :param data(bytearray): data to be written to disc
+        '''
+        storage = self._resource_storage.get(name)
+        storage.write(data)
+        storage.close()
+
+    def _write_schema(self, name):
+        '''
+        Writes resource schema
+
+        :param name: name of resource
+        '''
+        self._write_raw_data(f"{name}.schema", bytes(
+            self._RESOURCES[name].schema, 'utf-8'))
+
+    def _write_archive_signature(self):
+        '''Writes archive's signature'''
+        self._write_raw_data(f"{self._NAME}.archive", b'\x00' * 16)
+
+    def _write_archive_schema(self):
+        '''Writes archive schema'''
+        self._write_raw_data(
+            f"{self._NAME}.archive.schema", bytes(self._SCHEMA, 'utf-8'))
+
+    def _write_index_schema(self, resource_name, schema):
+        self._write_raw_data(
+            f"{resource_name}_index.schema", bytes(schema, 'utf-8'))
+
+    def subarchive(self, name):
+        """
+        Returns an archive builder for the sub-archive `name`.
+        :raises $name_not_subarchive_error
+        :param name: name of the sub-archive
+        """
+        NotImplemented
+
+    @classmethod
+    def validate_structure_fields(cls, name, struct, initializer):
+        '''
+        Validates whether passed object has all required fields
+
+        :raises MissingFieldError
+        :raises UnknownFieldError
+        :param name(str): name of object(struct)
+        :param struct(object): object to validate
+        :param initializer(object): provided field keys to validate from
+        '''
+        for key in initializer._FIELD_KEYS:
+            if key not in struct:
+                raise MissingFieldError(key, name)
+        for key in struct.keys():
+            if key not in initializer._FIELD_KEYS:
+                raise UnknownFieldError(key, name)
+
+    def set_instance(self, storage, name, value):
+        '''
+        Creates and writes instance type resource
+
+        :param storage(object): handles storage and writing to disc
+        :param name(str): instance name
+        :param value(dict): instance object replicates struct
+        '''
+        initializer = self._RESOURCES[name].initializer
+        ArchiveBuilder.validate_structure_fields(name, value, initializer)
+
+        bout = bytearray(initializer._SIZE_IN_BYTES)
+        for (key, field) in initializer._FIELDS.items():
+            write_value(bout, field.offset, field.width,
+                        field.is_signed, value[key])
+
+        storage.write(bout)
+
+    def set_vector(self, storage, name, vector):
```

Comment:
It is a bit weird to have these as public methods, even though they are not self-contained (e.g. they do not set the padding, size, etc).

I would either have them private, or make them self-contained (and if they are self-contained... do we even need a generic `set` method?)

Reply:
I will make them private as they are not supposed to be used outside of this class. Generic set function is also performing task of writing resource schema and other stuffs.